posts/2023-06-07-kotlin-ktor-keycloak.adoc

```diff
@@ -0,0 +1,357 @@
+----
+
+Besides creating the http client we also need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization,
+where we will ignore the json fields that we are not interested in.
+
+[source,kotlin]
+----
+    val client = HttpClient(CIO) {
+        install(ContentNegotiation) {
+            json(
+                Json {
+                    ignoreUnknownKeys = true
+                },
+            )
+        }
+    }
+----
+
+Now that the http client is properly setup we will start by retrieving a JWT token from Keycloak:
+
+[source,kotlin]
+----
+@Serializable
+data class BotToken(val access_token: String)
+
+val token = client.submitForm(
+        url = "http://localhost:8080/realms/lunarealm/protocol/openid-connect/token",
+        formParameters = parameters {
+            append("grant_type", "client_credentials")
+            append("client_id", "luna-vacation-bot")
+            append("client_secret", "<replace by client_secret>")
+        },
+    ).body<BotToken>().access_token
+----
+
+and then we can finally call the `luna-vacation-api`:
+
+[source,kotlin]
+----
+val response = client.get("http://localhost:5050/imionvacation") {
+        bearerAuth(token)
+    }.bodyAsText()
+----
+
+And what do we do with the `response` now? We send it to https://slack.com/[Slack] using a https://slack.com/apps[Slack App]. So let's do that next.
+
+
+=== Create a Slack app
+You can create your own workspace at https://slack.com/[slack.com]. Once you have it you can create a new https://api.slack.com/apps[App]. Choose `Create New App` and then `From scratch`:
+
+.Create a new Slack App
+image::slack_create_app_1.png[Diagram, 400]
+
+I named the app `ImIOnVacation` and added it to my personal workspace:
+
+.Configure a Slack App worskpace
+image::slack_create_app_2.png[Diagram, 450]
+
+To finalize the App creation you need to set some permissions:
+
+.Setup Slack App permissions
+image::slack_app_permissions.png[Diagram, 400]
+
+In `Basic Information` you can personalize your app appearance, but for now let's jump to `OAuth & Permission` and, from there, scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:
+
+.Setup User Token Scopes
+image::slack_app_scopes.png[Diagram, 500]
+
+Then scroll up and click `Install to Workspace`. That will take you to a permissions screen: click `Allow`. You can see that a `User OAuth Token` was created as well.
+
+.Slack App User OAuth token
+image::slack_user_oauth_token.png[Diagram, 500]
+
+We will need this `User OAuth Token` to be able to send messages to Slack via this App.
+
+=== Put it all together
+
+We can now send that `luna-vacation-api` response to a Slack channel. The channel I'll use is the chat with myself.
+In Slack, channel ids can be seen when you select a channel or a direct message chat
+and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".
```

Comment:
```suggestion
On Slack, channel IDs can be seen when you select a channel or a direct message chat
and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".
```

I think that this "when you select a channel or a direct message chat
and then click on the top details of the channel/person" is a bit confusing.

Reply:
I'm going to simplify it with `On Slack, channel IDs can be seen on the channel or chat details, at the very bottom.`